BTrees/tests/common.py

```diff
@@ -1595,6 +1596,54 @@ def test_legacy_py_pickle(self):
             self.assertTrue(b'Py' not in s2)
             self.assertEqual(s2, s)
 
+    def testSetstateBadChild(self):
+        # tree used to allow to pass in non (tree or bucket) node as a child
+        # via __setstate__. This was leading to crashes when using C mode.
+        t = self._makeOne()
+        b = t._bucket_type()
+        if isaset(b):
+            b.add(1)
+        else:
+            b[1] = 11
+
+        # xchild is non-BTree class deriving from Persistent
+        import persistent
+        xchild = persistent.Persistent()
+        self.assertIs(xchild._p_oid, None)
+
+        typeErrOK = "tree child %s is neither %s nor %s" % \
+                        (_tp_name(type(xchild)), _tp_name(type(t)),
+                         _tp_name(t._bucket_type))
+
+        # if the following is allowed, e.g.
+        # t.__getstate__(), or t[0]=1 corrupt memory and crash.
+        try:
+            t.__setstate__(
+                    (
+                        (xchild,), # child0 is neither tree nor bucket
+                        b
+                    )
+                )
+        except TypeError as v:
+            self.assertEqual(str(v), typeErrOK)
+        else:
+            #t[0]=1
+            raise AssertionError("Expected exception")
```

Comment:
```suggestion
with self.assertRaises(TypeError) as exc:
    t.__setstate__(
            (
                (xchild,), # child0 is neither tree nor bucket
                b
            )
        )
self.assertEqual(str(exc.exception), typeErrOK)
```

I think that will take care of the coverage reduction.